### PR TITLE
S3 refactor reconcile command

### PR DESCRIPTION
**Pull request by jonavellecuerdo**
### Purpose and background context
Update `reconcile` command to support file retrieval from S3.

### How can a reviewer manually see the effects of these changes?
Run `make test` and verify all test are passing 

### Includes new or updated dependencies?
NO

### Changes expectations for external applications?
NO

### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/ES-2719

### Developer
- [X] All new ENV is documented in README
- [ ] All new ENV has been added to staging and production environments
- [ ] All related Jira tickets are linked in commit message(s)
- [X] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed and verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

**Comment by jonavellecuerdo** [comment]
@ehanson8 I kept the description short given that we discussed a lot of the updates implemented in this PR during our sync yesterday. Here are some changes I made:

* Simplify `parse_identifier_from_file_name` to `parse_value_from_text`: Regex is also used to retrieve the bucket and prefix from a provided `s3_path` to the `reconcile` command as well, so I decided to extend this function to "getting a value from regex".
* Remove indexing (string splicing) option to extract file identifiers from S3 file object keys.

File: config/aspace.json [hunk 1]
```diff
@@ -0,0 +1,48 @@
+{
+    "settings": {
+        "search_in": [
+            "objects"
+        ],
+        "identifiers": {
+            "id_method": "regex",
+            "id_regex_pattern": ".*-(.*?-.*)\\..*$"
+        }
```
Comment (ehanson8) [inline config/aspace.json]: Looks like this needs to be updated based on what I'm seeing in `tests/fixtures/source_config.json`:

```
{
    "settings": {
        "object_folders": [
            "objects"
        ],
        "id_regex": ".*-(.*?-.*)\\..*$"
    },
```

And 👍 on  renaming to `object_folders`

File: config/test.json
Comment (ehanson8) [inline config/test.json:1]: Don't think we need this file, we can use the `tests/fixtures/source_config.json` for our test uploads

File: dsaps/cli.py [hunk 2]
```diff
@@ -74,9 +77,11 @@ def main(ctx, url, email, password):
         level=logging.INFO,
     )
     logger.info("Application start")
+    source_config = helpers.load_source_configs(config_file)
```
Comment (ehanson8) [inline dsaps/cli.py]: Rename to `load_source_config`

File: dsaps/cli.py [hunk 3]
```diff
@@ -74,9 +77,11 @@ def main(ctx, url, email, password):
...
     client = Client(url)
     client.authenticate(email, password)
     start_time = time.time()
+    ctx.obj["config"] = source_config
```
Comment (ehanson8) [inline dsaps/cli.py]: I think you could call `helpers.load_source_configs(config_file)` directly here instead of assigning it to a variable since it's not called elsewhere

File: dsaps/helpers.py [hunk 4]
```diff
@@ -11,21 +25,58 @@ def create_csv_from_list(list_name, output):
...
-def create_file_list(file_path, s3_client, file_type):
-    """Create a list of file names."""
-    files = []
+def get_files_from_s3(
+    s3_path: str,
+    s3_client,
+    id_regex: str,
+    object_folders=None,
+) -> dict:
+    """Get a list of files on S3 per unique file identifier.
```
Comment (ehanson8) [inline dsaps/helpers.py]: Change to `item identifier` instead of `file identifier`?

File: config/aspace.json [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    },
+    "mapping": {
+        "file_identifier": {
```
Comment (ehanson8) [inline config/aspace.json]: Update to `item_identifier` and we'll talk to DDC about changing how they format their CSVs

File: dsaps/helpers.py [hunk 6]
```diff
@@ -11,21 +25,58 @@ def create_csv_from_list(list_name, output):
...
+
+    Args:
+        s3_path (str): S3 path to the topmost level directory in which to look for files.
+            Formatted as "s3://bucket/prefix/" or "bucket/prefix/".
+        s3_client (_type_): S3 client.
+        id_regex (str): A regex expression for extracting the item identifier
+            from the S3 file object key.
+        object_folders (list[str], optional): A list of subfolders in which to look for files.
```
Comment (ehanson8) [inline dsaps/helpers.py]: I'd propose `bitstream_folders` to sync our vocab with DSpace

File: dsaps/helpers.py [hunk 7]
```diff
@@ -11,21 +25,58 @@ def create_csv_from_list(list_name, output):
...
+           The method will iterate over the list and check if any of the subfolders appear
+           in the file object key. Defaults to None.
```
Comment (ehanson8) [inline dsaps/helpers.py]: This is a little hard to parse, rephrase?

File: dsaps/helpers.py [hunk 8]
```diff
@@ -11,21 +25,58 @@ def create_csv_from_list(list_name, output):
...
+        for file in page["Contents"]:
+            file_path = file["Key"]
+            file_name = file_path.split("/")[-1]
+            if object_folders:
+                # if the object is not stored in any of the folders specified
+                # exclude object
+                if not [folder for folder in object_folders if folder in file_path]:
+                    continue
+            file_identifier = parse_value_from_text(file_name, id_regex)
+            files.setdefault(file_identifier, []).append(file["Key"])
```
Comment (ehanson8) [inline dsaps/helpers.py]: Update to `item_identifier`

File: dsaps/helpers.py [hunk 9]
```diff
@@ -11,21 +25,58 @@ def create_csv_from_list(list_name, output):
...
     return files
 
 
+def parse_value_from_text(
```
Comment (ehanson8) [inline dsaps/helpers.py:71]: Good call to extend this function!

File: tests/conftest.py [hunk 10]
```diff
@@ -19,6 +19,82 @@ def _test_environment(monkeypatch):
...
+        s3 = boto3.client("s3")
+        s3.create_bucket(Bucket=bucket_name)
+        yield bucket_name
+
+
+@pytest.fixture
+def mocked_s3_one_file_per_file_id(mocked_bucket):
+    s3 = boto3.client("s3")
+    s3.put_object(
+        Body=mocked_bucket,
```
Comment (ehanson8) [inline tests/conftest.py]: As discussed, don't think `Body` should be `mocked_bucket`

File: tests/test_helpers.py [hunk 11]
```diff
@@ -1,31 +1,135 @@
...
+def test_get_files_from_s3_ddc_one_file_per_file_id(
+    mocked_s3_ddc_one_file_per_file_id, s3_client
+):
+    files = get_files_from_s3(
+        s3_path="s3://mocked-ddc-bucket/subfolder",
+        s3_client=s3_client,
+        object_folders=["objects"],
+        id_regex=REGEX_ID_DDC,
+    )
+    print(files)
```
Comment (ehanson8) [inline tests/test_helpers.py]: Remove `print`

File: tests/test_helpers.py [hunk 12]
```diff
@@ -36,34 +140,37 @@ def test_create_ingest_report(runner, output_dir):
...
     assert len(file_matches) == 1
     assert "test" in file_matches
 
 
 def test_update_metadata_csv(output_dir):
     """Test update_metadata_csv function."""
     metadata_matches = ["test"]
-    helpers.update_metadata_csv(
-        "tests/fixtures/aspace_metadata_delimited.csv", output_dir, metadata_matches
+    update_metadata_csv(
```
Comment (ehanson8) [inline tests/test_helpers.py:162]: Let's update this test to check the `bitstreams` column as well

File: tests/conftest.py [hunk 13]
```diff
@@ -19,6 +19,82 @@ def _test_environment(monkeypatch):
...
+    s3.put_object(
+        Body="",
+        Bucket=mocked_bucket,
+        Key="many-to-one/bbbb_003_01.jpg",
+    )
+    return mocked_bucket
+
+
+@pytest.fixture
+def mocked_s3_ddc_one_file_per_file_id():
```
Comment (ehanson8) [inline tests/conftest.py]: What's difference between this and `mocked_s3_one_file_per_file_id`? Could we get away with just one?

File: tests/conftest.py [hunk 14]
```diff
@@ -19,6 +19,82 @@ def _test_environment(monkeypatch):
     monkeypatch.setenv("AWS_SESSION_TOKEN", "testing")
 
 
+@pytest.fixture
```
Comment (ehanson8) [inline tests/conftest.py:22]: Are all 3 of these bucket fixtures necessary? Couldn't they all be on a single mocked bucket with different paths?

**Review by ehanson8 — CHANGES_REQUESTED** [review]
Looks great overall but some requested changes!

File: dsaps/helpers.py (hunk 7 above)
Comment (jonavellecuerdo) [reply dsaps/helpers.py]: Let me know what you think of the updated language!

File: tests/conftest.py (hunk 10 above)
Comment (jonavellecuerdo) [reply tests/conftest.py]: 😅 Thanks for the reminder!

File: tests/conftest.py (hunk 14 above)
Comment (jonavellecuerdo) [reply tests/conftest.py:22]: That is a very good point and it also is a stronger test as it would confirm the `s3_path` parameter is parsed correctly to retrieve files from the correct location.

File: tests/conftest.py (hunk 13 above)
Comment (jonavellecuerdo) [reply tests/conftest.py]: I suppose there isn't. I think the test I'm actually looking for is making sure that `parse_value_from_text()` can successfully get the item identifier from the S3 file object key given the regex specified in the config. 🤔 

Whenever we get a new source, I think we'll need to do two things:
* Add a source config JSON to `config/`
* Add a test to confirm the regex specified in the config successfully retrieves the item identifier

I updated the tests:
* [Verify regex for DDC works ](https://github.com/MITLibraries/dspace-api-python-scripts/pull/38/files#diff-08a6944653efc734f4c8cb856fc854b9f21d655ca69796af2b170c53bff7919eR50-R58):
   * Looking at the files Joe sent, I see two file name formats: 
      a. [4336711c-c826-46cd-a39c-1739450932eb-02-000463360.pdf](https://s3.console.aws.amazon.com/s3/object/cdps-storage-stage-dissemination?region=us-east-1&bucketType=general&prefix=dipstore2/3e4b/7286/1d74/43f4/ac0a/cbd4/a859/f1ba/2024_028RR_002-dc01295e-5248-4770-93aa-9c22b2f2e07c/objects/4336711c-c826-46cd-a39c-1739450932eb-02-000463360.pdf)
      b. [0a3b7ebd-c819-40a5-90ed-f941877c0f4b-02-000065980_005.jpg](https://s3.console.aws.amazon.com/s3/object/cdps-storage-stage-dissemination?region=us-east-1&bucketType=general&prefix=dipstore2/6ffa/30cc/1dd9/4da9/8480/fa26/8cbe/9c05/2023_068RR_001-aa2a33c4-4222-42b8-98d4-c44e5831f3fd/objects/0a3b7ebd-c819-40a5-90ed-f941877c0f4b-02-000065980_005.jpg)

     I had to make a small update to [REGEX_ID_DDC](https://github.com/MITLibraries/dspace-api-python-scripts/pull/38/files#diff-08a6944653efc734f4c8cb856fc854b9f21d655ca69796af2b170c53bff7919eR23) to exclude everything after the underscore from file names formatted like **b.**

* [Add a test to verify the `bitstream_folders` parameter works](https://github.com/MITLibraries/dspace-api-python-scripts/pull/38/files#diff-08a6944653efc734f4c8cb856fc854b9f21d655ca69796af2b170c53bff7919eR115-R125)
   **Note**: A [mocked bucket fixture](https://github.com/MITLibraries/dspace-api-python-scripts/pull/38/files#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128R58-R68) was created for this test. Currently the config JSON is specific to (a) source and (b) S3 bucket structure. Might be nice update this in the future, but I think it is beyond the scope of this PR. Let me know what you think!

File: tests/test_helpers.py (hunk 12 above)
Comment (jonavellecuerdo) [reply tests/test_helpers.py:162]: Hmm, can you clarify this request? 🤔

**Comment by jonavellecuerdo** [comment]
@ehanson8 Thank you for your review! I have one question for ya'!

File: tests/test_helpers.py (hunk 12 above)
Comment (ehanson8) [reply tests/test_helpers.py:162]: Checking the `bitstreams` column that is now added by `update_metadata_csv` to make sure it has the expected values, happy to chat about it if needed!

Comment (jonavellecuerdo) [reply tests/test_helpers.py:162]: I understand now! I just pushed one commit related to this.

**Review by ehanson8 — APPROVED** [review]
